### ISSUE TITLE
[BUG] Very Serious Bug When using "metric=precomputed".

### ISSUE THREAD

**Issue by Wall-ee** [lmcinnes/umap#864]
First of all,  I want to calc a very large  dataset using umap to make dimension  reduction. Considering the document that a precomputed matrix could accelerate the computing speed.

So, I used "metric=precomputed" switch to perform a better speed. 
However， I  found the result is  completely different from the  original  method ！！！
I've  already located the problem,  here is a introduction.
So  example  code:

```
import umap
import scipy 
import numpy as np
from sklearn.metrics import pairwise_distances

#version for  original code 
rd = np.random.RandomState(888) 
aa = rd.normal(5, 1, [40,50])
model1 = umap.UMAP(random_state=10)
embedding = model1.fit_transform(aa)

print(embedding)

#using precomputed distance
bb = pairwise_distances(aa)
from scipy.sparse import csr_matrix
cc = csr_matrix(bb)

model3  = umap.UMAP(metric='precomputed',random_state=10)
embedding2 = model3.fit_transform(cc)
print(embedding2)

```
There  is  the first problem, If a precomputed matrix needed to use, the matrix should be a sparse matrix!!! other wise the code will calc the dist matrix again!!

umap.py  row=2364
```
if self.metric == "precomputed" and self._sparse_data:
    # For sparse precomputed distance matrices, we just argsort the rows to find
    # nearest neighbors. To make this easier, we expect matrices that are
    # symmetrical (so we can find neighbors by looking at rows in isolation,
    # rather than also having to consider that sample's column too).
    # print("Computing KNNs for sparse precomputed distances...")
    if sparse_tril(X).getnnz() != sparse_triu(X).getnnz():
```

so the result from above side is :

```
#embedding 1
[[-6.777242  -2.9653606]
 [-5.312394  -3.720777 ]
 [-7.0402703 -4.865579 ]
 [-7.494709  -4.1032014]
 [-6.739107  -4.2637396]
 [-6.359506  -4.032858 ]
    .......

#embedding 2
  [[-8.083555  15.876725 ]
 [-8.736375  14.2785635]
 [-6.570646  15.637847 ]
 [-6.1307974 14.759017 ]
 [-6.2153373 15.248924 ]
 [-6.284779  16.172754 ]
 ......


```

So, I tracked the original code to find what cause the difference.

then a find here !

umap.py  row=562 in the function of "fuzzy_simplicial_set"

this is a core function of umap

```

knn_indices, knn_dists, _ = nearest_neighbors(
    X,
    n_neighbors,
    metric,
    metric_kwds,
    angular,
    random_state,
    verbose=verbose,
)

```

the bug is IF X is a sparse style matrix, the calc result is different from  when X is a dense matrix!!!!!!

result 

```
# when x is sparse matrix
# x
<40x40 sparse matrix of type '<class 'numpy.float32'>'
	with 1560 stored elements in Compressed Sparse Row format>
X.shape
Out[3]: (40, 40)

knn_dists
[[ 7.33820343,  7.59337425,  8.05820179,  8.16062737,  8.27206707,
         8.3769598 ,  8.52959538,  8.54904938,  8.54981995,  8.65421963,
         8.75841427,  8.81221962,  8.81944847,  8.87848377,  8.88055706],
       [ 9.24660969,  9.44158745,  9.50287247,  9.5380373 ,  9.75831985,
         9.76912975,  9.88988495,  9.95298195, 10.00327873, 10.03878117,
        10.13149166, 10.21953869, 10.27251053, 10.27710056, 10.33901501],
.......

# we convert X to dense matrix 
# using the same function again and the reset of the parameter are all the same

aa = X.todense()

a, b, _ = nearest_neighbors(
    aa,
    n_neighbors,
    metric,
    metric_kwds,
    angular,
    random_state,
    verbose=verbose,
)

# show b
b
matrix([[ 0.       ,  7.3382034,  7.5933743,  8.058202 ,  8.160627 ,
          8.272067 ,  8.37696  ,  8.529595 ,  8.549049 ,  8.54982  ,
          8.65422  ,  8.758414 ,  8.81222  ,  8.819448 ,  8.878484 ],
        [ 0.       ,  9.24661  ,  9.441587 ,  9.502872 ,  9.538037 ,
          9.75832  ,  9.76913  ,  9.889885 ,  9.952982 , 10.003279 ,
         10.038781 , 10.131492 , 10.219539 , 10.272511 , 10.277101 ],
        [ 0.       ,  8.134261 ,  8.294831 ,  8.376041 ,  8.377901 ,
          8.489849 ,  8.676778 ,  8.7918825,  8.965539 ,  9.050419 ,
          9.153272 ,  9.177004 ,  9.195522 ,  9.284557 ,  9.451907 ],
    .......

```

Which means the same function of  nearest_neighbors will generate differnt output when X  is dense and X is sparse matrix!!

umap.py row= 254

```
def nearest_neighbors(
    X,
    n_neighbors,
    metric,
    metric_kwds,
    angular,
    random_state,
    low_memory=True,
    use_pynndescent=True,
    n_jobs=-1,
    verbose=False,
):

```

this is a very serious problem~~~

please fix it 

my package version

umap.__version__ == '0.5.3'
